[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108434
* __->__ #108298

The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

fegin committed:
[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True

The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin commented:
@awgu Can you review if the implementation of concat with aligned padding is correct?

wz337 commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -142,8 +142,8 @@ def forward(self, x):
         super().__init__()
-        self.block0 = BlockB(5, 7)
-        self.block1 = BlockB(7, 7)
+        self.block0 = BlockB(5, 3)
```
Noob question: what's the difference between the two?

fegin replied:
With this new setting, we were able to reproduce the align padding issue.

awgu replied:
This does make me wonder if it would be worth utility to visualize/print the padding.

fegin replied:
Yup, since we are going to maintain FSDP FlatParameter version for a while, having simple utility to print the padding is going to be useful. We can make it a ramp-up task.

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
 
 
+@dataclass
+class StateInfo:
```
nit: Do you think it is worth documenting that keys are state names, e.g. `"exp_avg"`?

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+    shard_state: bool,
+) -> Dict[str, Any]:
+    fsdp_osd_state: Dict[str, Any] = {}
+    all_states: Dict[int, Dict[str, Any]] = {}
```
Note to self: The key is only used to de-dup against `FSDPParamInfo` objects. One `FSDPParamInfo` object corresponds to one `FlatParamHandle`/`FlatParameter`.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+            continue
+
+        if optim_state_key.is_fsdp_managed:
+            fqn = optim_state_key.unflat_param_names[0]
```
Note to self: For `use_orig_params=True` path, it seems that `unflat_param_names` is always a singleton tuple.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+                if padding_len:
+                    local_buffers.append(empty_func(padding_len))
+            else:
+                if buffers[buffer_idx] is not None:
```
When is this `False`?

fegin replied:
When the rank does not own the buffer. Add a comment to explain.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+) -> None:
+    """
+    Given a output state dict, ``output_states``, which the keys are FQNs to the
+    original parameters (not FlatParameters nor parmeter ID), and the values
```
nit: 
```suggestion
    original parameters (not FlatParameters nor parameter ID), and the values
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+            # Add right-handed padding.
+            local_buffers.append(empty_func(shard_numel_padded))
+        local_shard = torch.cat(local_buffers)
+        assert local_shard.numel() * fsdp_state.world_size == gathered_tensor.numel()
```
nit: It looks like if this `assert` fails, then there is a bug inside `flat_param.py`'s metadata.
For completeness, should we add an assert message with the relevant info?

fegin replied:
It may not be a bug of FlatParameter's metadata but some inconsistency of the reconstruction logic here. Maybe some  edge cases. Added the message.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+        unpadded_tensor = gathered_tensor[: flat_param._unpadded_unsharded_size.numel()]
+        flat_param_handle = fsdp_param_info.handle
+        orig_states = flat_param_handle._get_unflat_views_aligned(unpadded_tensor)
+        assert len(orig_states) == len(fsdp_param_info.param_indices)
```
nit: Similarly, maybe just include the `assert` message in case?

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+    Given a optimizer state dict, ``input_states``, which the keys are FQNs to the
+    original parameters (not FlatParameters nor parmeter ID), gather all the
+    states and unflatten them to the original dimensions. Note that all the
+    params refered by the ``input_states`` must be manageded by FSDP.
```
nit: 
```suggestion
    params referred by the ``input_states`` must be managed by FSDP.
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+                    # align padding.
+                    padding_len = (
+                        padding_end - begin + 1
+                        if end >= padding_end
```
Note to self: `end < padding_end` means that this rank's entire sharded flat parameter is padding since then `padding_begin <= begin < end < padding_end`.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
+                    # align padding.
+                    padding_len = (
+                        end - padding_begin + 1
+                        if begin <= padding_begin
```
Note to self: Similarly, if `begin > padding_begin`, then `padding_begin < begin < end <= padding_end`, and the rank's entire sharded flat parameter is padding.

awgu reviewed (approved):
The padding logic looks good to me!

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True"


The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
